File: 768-partition-labels/partition-labels.py
class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        cache = {}
        for i, c in enumerate(s):
            cache[c] = i
        result = []
        curr = end = 0
        for i, c in enumerate(s):
            curr += 1
            end = max(end, cache[c])
            if i == end:
                result.append(curr)
                curr = 0
        return result

File: 768-partition-labels/partition-labels.java
class Solution {
    public List<Integer> partitionLabels(String s) {
        HashMap<Character, Integer> cache = new HashMap<Character, Integer>();
        for (int i = 0; i < s.length(); i++ ) cache.put(s.charAt(i), i);

        List<Integer> result = new ArrayList<Integer>();
        int step = 0, end = 0;
        for (int i = 0; i < s.length(); i++ ){
            step += 1;
            end = Math.max(end, cache.get(s.charAt(i)));
            if(i == end){
                result.add(step);
                step = 0;
            }
        }
        return result;
    }
}